src/mmap.rs

```diff
@@ -0,0 +1,585 @@
+    ///   let _: u32 = gm.read_obj(read_addr).unwrap();
+    /// ```
+    fn write_from_stream<F>(
+        &self,
+        addr: MemoryRegionAddress,
+        src: &mut F,
+        count: usize,
+    ) -> Result<()>
+    where
+        F: Read,
+    {
+        let maddr = addr.raw_value() as usize;
+        self.as_volatile_slice()
+            .write_from_stream::<F>(maddr, src, count)
+            .map_err(Into::into)
+    }
+
+    /// Reads data from the region to a writable object.
+    ///
+    /// # Examples
+    ///
+    /// * Write 128 bytes to /dev/null
+    ///
+    /// ```
+    /// # use vm_memory::{Address, Bytes, GuestAddress, GuestMemoryMmap};
+    /// # use std::fs::OpenOptions;
+    /// # let start_addr = GuestAddress(0x1000);
+    /// # let gm = GuestMemoryMmap::new(&vec![(start_addr, 0x400)]).unwrap();
+    ///   let mut file = OpenOptions::new().write(true).open("/dev/null").unwrap();
+    ///   let mut mem = [0u8; 1024];
+    ///   gm.read_into_stream(start_addr, &mut file, 128).unwrap();
+    /// ```
+    fn read_into_stream<F>(
+        &self,
+        addr: MemoryRegionAddress,
+        dst: &mut F,
+        count: usize,
+    ) -> Result<()>
+    where
+        F: Write,
+    {
+        let maddr = addr.raw_value() as usize;
+        self.as_volatile_slice()
+            .read_into_stream::<F>(maddr, dst, count)
+            .map_err(Into::into)
+    }
+}
+
+impl GuestMemoryRegion for GuestRegionMmap {
+    fn len(&self) -> GuestAddressValue {
+        self.mapping.len() as GuestAddressValue
+    }
+
+    fn min_addr(&self) -> GuestAddress {
+        self.guest_base
+    }
+
+    unsafe fn as_slice(&self) -> Option<&[u8]> {
+        Some(self.mapping.as_slice())
+    }
+
+    unsafe fn as_mut_slice(&self) -> Option<&mut [u8]> {
+        Some(self.mapping.as_mut_slice())
+    }
+}
+
+/// Tracks memory regions allocated/mapped for the guest in the current process.
+#[derive(Clone)]
+pub struct GuestMemoryMmap {
+    regions: Arc<Vec<GuestRegionMmap>>,
+}
+
+impl GuestMemoryMmap {
+    /// Creates a container and allocates anonymous memory for guest memory regions.
+    /// Valid memory regions are specified as a Vec of (Address, Size) tuples sorted by Address.
+    pub fn new(ranges: &[(GuestAddress, usize)]) -> std::result::Result<Self, MmapError> {
+        if ranges.is_empty() {
+            return Err(MmapError::NoMemoryRegion);
+        }
+
+        let mut regions = Vec::<GuestRegionMmap>::new();
+        for range in ranges.iter() {
+            if let Some(last) = regions.last() {
+                if last
+                    .guest_base
+                    .checked_add(last.mapping.len() as GuestAddressValue)
+                    .map_or(true, |a| a > range.0)
+                {
+                    return Err(MmapError::MemoryRegionOverlap);
+                }
+            }
+
+            // seems like a false warning
+            #[allow(clippy::redundant_closure)]
```

Comment:
Hmm.. This is strange. It should work like:
```rust
let mapping = MmapRegion::new(range.1).map_err(MmapError::SystemCallFailed)?;
```
What rust version are you using? Maybe it is something related to the version?
I wrote a minimal example in [Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=27f4da2774fd8aaef232b5b81a5d0534).

Reply:
yah, it works. Haven't realized that rust has sugar for this case.
Will wait for more comments and fix this in coming patches.